### PR TITLE
[Merged by Bors] - Add invalid post malfeasance

### LINKED ISSUE #6141: Malfeasance V2: Invalid Post
## Description

Contextual validation includes a check that verifies that some labels of the provided PoST of an identity are valid. If those labels are found to not be valid a malfeasance proof needs to be created that proofs to other nodes that the identity in question produced an invalid PoST, since those might not validate the same labels and might consider the ATX valid.

## Implementation hints

The malfeasance proof needs to include the following data to be considered valid:

1. The PoST that is invalid - the receiver of the malfeasance proof should do a full validation of the PoST so we don't need to keep track of which label is invalid
2. A proof about the commitment ATX for that identity, i.e. the merkle proof to it in the initial ATX of that identity
3. In the case that the PoST was created by an identity that is part of a marriage set information about this marriage set as far as it is needed to validate the post.

### PR DESCRIPTION
**Pull request by fasmat**
## Motivation

Closes #6141. This PR adds the Malfeasance proof needed to show that an identity produced an invalid post. Supports both solo and merged ATXs.

## Description

This PR adds some additional methods to the ATX wire types to deduplicate code that is needed to create and validate malfeasance proofs. This also reduces the code needed for the `ProofDoubleMarry` and `ProofInvalidPost` proofs significantly.  I will also re-base #6339 and #6310 to use the new helpers in the wire package and reduce code duplication.

The interface for `ATXProof.Proof` has been updated to use a `MalfeasanceValidator` instead of a `*signing.EdVerifier`, since more than just signatures are validated with some malfeasance proofs. The new interface is defined in the `wire` package and the current implementation is part of the `activation.Validator` type but will be moved to the malfeasance handler and/or publisher in #6307 (which will get rid of the open TODO there).

**IMPORTANT:** ID calculation for ATXv2 changed - when we merge this we cannot just update atxmerge testnets without causing problems. This change was necessary because some leaves in the ATXID tree were not `Hash32` which still lets one calculate the root of the tree, but causes problems when trying to validate a merkle proof against it.

`atxTreeHash` now uses the `blake3.Hash` pool that we have, before it initialized a new hasher on every call which was expensive and is not needed if an existing instance can be reused from the pool, this more than tripled execution time of the tests in the `activation/wire` package on my machine.

## Test Plan

Several unit tests were added to test the functionality of the invalid post proof.

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed
- [x] Update [changelog](../CHANGELOG.md) as needed

### CODE REVIEW

File: activation/handler_v2.go [hunk 1]
```diff
@@ -674,6 +667,49 @@ func (h *HandlerV2) syntacticallyValidateDeps(
...
+
+		var initialAtxBytes sql.Blob
+		v, err := atxs.LoadBlob(ctx, h.cdb, initialID.Bytes(), &initialAtxBytes)
+		if err != nil {
+			return fmt.Errorf("fetch initial ATX blob for ID %s: %w", nodeID.ShortString(), err)
+		}
+		if v != types.AtxV2 {
+			// TODO(mafa): this needs to be fixed
+			return fmt.Errorf("initial ATX is not V2 for ID %s", nodeID.ShortString())
+		}
```
Comment (fasmat) [inline activation/handler_v2.go:699]: TODO (mafa): fix this

File: activation/handler_v2_test.go [hunk 2]
```diff
@@ -1523,6 +1523,7 @@ func TestHandlerV2_SyntacticallyValidateDeps(t *testing.T) {
 		require.ErrorContains(t, err, "post failure")
 	})
 	t.Run("invalid PoST index - generates a malfeasance proof", func(t *testing.T) {
+		// TODO(mafa): add such a test for solo and merged ATXs
```
Comment (fasmat) [inline activation/handler_v2_test.go:1526]: TODO (mafa): add these 2 tests

File: activation/validation.go [hunk 3]
```diff
@@ -236,6 +248,19 @@ func (v *Validator) PostV2(
...
+	nodeId types.NodeID,
+	commitmentAtxId types.ATXID,
+	post *types.Post,
+	challenge []byte,
+	numUnits uint32,
+	idx int,
+) error {
+	// TODO(mafa): this should be in the malfeasance handler with a dependency on the validator
+	return v.PostV2(ctx, nodeId, commitmentAtxId, post, challenge, numUnits, PostIndex(idx))
+}
```
Comment (fasmat) [inline activation/validation.go:262]: TODO(mafa): add this to #6307

File: activation/handler_v2.go [hunk 4]
```diff
@@ -674,6 +667,44 @@ func (h *HandlerV2) syntacticallyValidateDeps(
...
+		initialID, err := atxs.GetFirstIDByNodeID(h.cdb, nodeID)
+		if err != nil {
+			return fmt.Errorf("fetch initial ATX for ID %s: %w", nodeID.ShortString(), err)
+		}
+
+		// TODO(mafa): what if initial ATX is not v2?
+		initialAtx, err = h.fetchWireAtx(ctx, h.cdb, initialID)
+		if err != nil {
+			return fmt.Errorf("fetch initial ATX blob for ID %s: %w", nodeID.ShortString(), err)
+		}
```
Comment (fasmat) [inline activation/handler_v2.go:695]: TODO: Needs to be addressed in a followup PR.

**Review by jellonek — APPROVED** [review]

**Comment by fasmat** [comment]
bors merge

**Comment by fasmat** [comment]
bors merge

**Comment by fasmat** [comment]
TestPartition_30_70 failed because 4 pods were re-deployed shortly before the end of the test

bors merge

**Comment by fasmat** [comment]
bors merge

**Pull request closed without merging**